app/models/git_repository.rb

```diff
@@ -114,6 +114,27 @@ def file_changed?(sha1, sha2, file)
     executor.execute!("cd #{pwd}", "git diff --quiet --name-only #{sha1}..#{sha2} #{file}")
   end
 
+  def directory_contents(git_ref, path: '.')
+    cmd = "git ls-tree -r #{git_ref} --name-only --full-tree #{path}"
+    valid, output = run_single_command(cmd) { |line|
+      if block_given?
+        yield line.chomp
+      else
+        line.chomp
+      end
+    }
+    Rails.logger.error("Could not list the files in directory '#{path}'. Git Ref '#{git_ref}'") unless valid
+    output
+  end
+
+  def file_contents(git_ref, file_path)
+    #first command disables less as default pager
+    cmds = ["git config core.pager ''", "git show -q #{git_ref}:#{file_path}"]
+    valid, output = run_commands(cmds) { |line| line }
+    Rails.logger.error("Could not retrieve file contents for '#{file_path}'. Git Ref '#{git_ref}'") unless valid
+    output.join
+  end
```

Comment:
One thing you're leaving out in this workflow is a `git fetch`.  There's no guarantee that the local cache of the repo is going to have the reference that's specified. So before calling `#file_contents`, you should call `#update!`.

So let's say you're fetching the values of 3 different files.  Using the logic in `Build#file_from_repo` would make 3 calls to GitHub, and cache the results in memcache for 1 hour.  If you use this logic, you'd do a single `git fetch` to bring the local repo up to date, and then run 3 local git commands to get the file contents.

I don't know off-hand which would be faster.  My guess is that `git fetch` can take some time, even if the repo is up to date.  But we could do some testing.